src/SymfonyCache/ReverseProxyTtlListener.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+namespace FOS\HttpCache\SymfonyCache;
+
+use FOS\HttpCache\SymfonyCache\CacheEvent;
+use FOS\HttpCache\SymfonyCache\Events;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Custom TTL handler for the symfony built-in HttpCache.
+ *
+ * This allows to use a custom header to control time to live in HttpCache and
+ * keep s-maxage for 3rd party proxies.
+ *
+ * @author David Buchmann <mail@davidbu.ch>
+ *
+ * {@inheritdoc}
+ */
+class ReverseProxyTtlListener implements EventSubscriberInterface
+{
+    private $ttlHeader;
+
+    const SMAXAGE_BACKUP = 'FOS-Smaxage-Backup';
+
+    public function __construct($ttlHeader = 'X-Reverse-Proxy-TTL')
+    {
+        $this->ttlHeader = $ttlHeader;
+    }
+
+    /**
+     * Use the TTL from the custom header rather than the default one.
+     *
+     * If there is such a header, the original s_maxage is backed up to the
+     * static::SMAXAGE_BACKUP header.
+     *
+     * @param CacheEvent $e
+     */
+    public function useCustomTtl(CacheEvent $e)
+    {
+        $response = $e->getResponse();
+        if (!$response->headers->has($this->ttlHeader)) {
+            return;
+        }
+        $backup = $response->headers->hasCacheControlDirective('s-maxage')
+            ? $response->headers->getCacheControlDirective('s-maxage')
+            : 'false'
```

Comment:
no, because then we would lose the `s-maxage: 0` case and consider that as s-maxage not set